tendermint/src/lite_impl/signed_header.rs

```diff
@@ -52,42 +45,109 @@ impl lite::Commit for block::signed_header::SignedHeader {
-                        let reason = format!(
+// this private helper function does *not* do any validation but extracts
+// all non-BlockIDFlagAbsent votes from the commit:
+fn non_absent_votes(commit: &block::Commit) -> Vec<vote::Vote> {
+    let mut votes: Vec<vote::Vote> = Default::default();
+    for (i, commit_sig) in commit.signatures.iter().enumerate() {
+        if commit_sig.is_absent() {
+            continue;
+        }
+
+        if let Some(val_addr) = commit_sig.validator_address {
+            if let Some(sig) = commit_sig.signature.clone() {
+                let vote = vote::Vote {
+                    vote_type: vote::Type::Precommit,
+                    height: commit.height,
+                    round: commit.round,
+                    block_id: Option::from(commit.block_id.clone()),
+                    timestamp: commit_sig.timestamp,
+                    validator_address: val_addr,
+                    validator_index: u64::try_from(i)
+                        .expect("usize to u64 conversion failed for validator index"),
+                    signature: sig,
+                };
+                votes.push(vote);
+            }
+        }
+    }
+    votes
+}
+
+// TODO: consider moving this into commit_sig.rs instead and making it pub
+impl block::commit_sig::CommitSig {
+    fn validate(&self, vals: &Set) -> Result<(), Error> {
+        match self.block_id_flag {
+            BlockIDFlag::BlockIDFlagAbsent => {
+                if self.validator_address.is_some() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "validator address is present for absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                if self.signature.is_some() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "signature is present for absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                // TODO: deal with Time
+                // see https://github.com/informalsystems/tendermint-rs/pull/196#discussion_r401027989
+            }
+            BlockIDFlag::BlockIDFlagCommit | BlockIDFlag::BlockIDFlagNil => {
+                if self.validator_address.is_none() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "missing validator address for non-absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                if self.signature.is_none() {
+                    fail!(
+                        Kind::ImplementationSpecific,
+                        "missing signature for non-absent CommitSig {:#?}",
+                        self
+                    );
+                }
+                // TODO: this last check is only necessary if we do full verification (2/3) but the
```

Comment:
seems to me this should not be the concern of Commit to validate if it's a part of some validator set, but it should be rather the job of external object, who's probably holding a validator set to make sure commit signatures match validator set. that's how it's done in Go. It's interesting that you've decided to revert this relationship.